libshortfin/src/shortfin/local/async.h

```diff
@@ -64,6 +64,105 @@ class SHORTFIN_API CompletionEvent {
   std::any resource_baton_;
 };
 
+// Object that will eventually be set to some completion state, either a result
+// value or an exception status. Like CompletionEvents, Futures are copyable,
+// and all such copies share the same state.
+class SHORTFIN_API Future {
+ public:
+  Future(const Future &other) = delete;
+  Future(Future &&other) = delete;
+  Future &operator=(const Future &other) = delete;
+  virtual ~Future();
+
+  void set_failure(iree_status_t failure_status) {
+    state_->failure_status_ = failure_status;
+    state_->done_ = true;
+    state_->done_event_.set();
+  }
+
+  // Returns whether this future is done.
+  bool is_done() { return state_->done_; }
+  bool is_failure() { return !iree_status_is_ok(state_->failure_status_); }
+  void ThrowFailure() {
+    if (!state_->done_) {
+      throw std::logic_error("Cannot get result from Future that is not done");
+    }
+    SHORTFIN_THROW_IF_ERROR(state_->failure_status_);
+  }
+
+  // Access the raw done wait source.
+  operator const iree_wait_source_t() { return state_->done_event_.await(); }
+
+ protected:
+  struct BaseState {
+    BaseState() : done_event_(false) {}
+    virtual ~BaseState() = default;
+    int ref_count_ = 1;
+    iree::event done_event_;
+    iree_status_t failure_status_ = iree_ok_status();
```

Comment:
it may be worth it to make a C++ wrapper for status - I found it useful - I'm not sure how many statuses (if any) are being leaked here. e.g. it may be a contractual requirement that ThrowFailure is always called before a Future ever goes out of scope, but I don't think that's guaranteed here and if failure_status_ points to a non-ok status then it'll leak.

all status instances must have iree_status_ignore or iree_status_free called on them - they both do the same thing but ignore improves readability when the user is explicitly ignoring an error (and we may have a debug flag at some point that prints out ignored errors to help track down missing ones). `runtime/src/iree/base/status_cc.h` is the one we provide in IREE that works like abseil Status (mostly), supports move semantics for transferring statuses, etc, and also has StatusOr if you need it.

Reply:
Thanks, did this. The use cases for wanting an ignorable status are pretty small, mainly in this kind of Future object where it is legal to completely ignore it from the user side.